dissect/target/tools/shell.py

```diff
@@ -504,20 +512,61 @@ def cmd_ls(self, args: argparse.Namespace, stdout) -> Optional[bool]:
             contents = [(path, path.name)]
 
         if not args.l:
-            print("\n".join([name for _, name in contents]), file=stdout)
+            for target_path, name in contents:
+                print(name, file=stdout)
+                if args.recursive and target_path.is_dir():
+                    dir_list.append(target_path)
+        else:
+            if len(contents) > 1:
+                print(f"total {len(contents)}", file=stdout)
+            for target_path, name in contents:
+                self.print_extensive_file_stat(args=args, stdout=stdout, target_path=target_path, name=name)
+                if args.recursive and target_path.is_dir():
+                    dir_list.append(target_path)
+
+        if args.recursive and len(dir_list):
+            for dir_name in dir_list:
+                self._recursive_ls(args, dir_name, stdout)
+
+    def _recursive_ls(self, args: argparse.Namespace, dir_name: fsutil.TargetPath, stdout: TextIO) -> Optional[bool]:
```

Comment:
There's currently quite a bit of code duplication between this and `cmd_ls`. Do you think you could re-use more? I.e. let `_recursive_ls` be the main body (perhaps rename it something like `_print_ls`?), and do the print on line 540 conditionally if the function is being called recursively (through an additional argument).